psi4/driver/qcdb/align.py

```diff
@@ -0,0 +1,737 @@
+    """Facilitates the application of the simple transformation operations
+    defined by namedtuple of arrays as recipe to the data structures
+    describing Cartesian molecular coordinates. Attaches functions to
+    transform the geometry, element list, gradient, etc. to the
+    AlignmentRecipe. When `mirror` attribute (defaults to False) active,
+    then molecular system can be substantively changed by procedure.
+
+    """
+
+    def __new__(cls, shift, rotation, atommap, mirror=False):
+        return super(AlignmentMill, cls).__new__(cls, shift, rotation, atommap, mirror)
+
+    def __str__(self, label=''):
+        width = 40
+        text = []
+        text.append('-' * width)
+        text.append('{:^{width}}'.format('AlignmentMill', width=width))
+        if label:
+            text.append('{:^{width}}'.format(label))
+        text.append('-' * width)
+        text.append('Mirror:   {}'.format(self.mirror))
+        text.append('Atom Map: {}'.format(self.atommap))
+        text.append('Shift:    {}'.format(self.shift))
+        text.append('Rotation:')
+        text.append('{}'.format(self.rotation))
+        text.append('-' * width)
+        return ('\n'.join(text))
+
+    def align_coordinates(self, geom, reverse=False):
+        """suitable for geometry or displaced geometry"""
+
+        algeom = np.copy(geom)
+        if reverse:
+            algeom = algeom.dot(self.rotation)
+            algeom = algeom + self.shift
+            if self.mirror:
+                algeom[:, 1] *= -1.
+        else:
+            if self.mirror:
+                algeom[:, 1] *= -1.
+            algeom = algeom - self.shift
+            algeom = algeom.dot(self.rotation)
+        algeom = algeom[self.atommap, :]
+        # mirror-wise, rsm/msr == rms/msr
+
+        return algeom
+
+    def align_atoms(self, ats):
+        """suitable for masses, symbols, Zs, etc."""
+
+        return ats[self.atommap]
+
+    def align_vector(self, vec):
+        """suitable for vector attached to molecule"""
+
+        # sensible? TODO
+        #alvec = np.copy(vec)
+        #if self.mirror:
+        #    alvec[:, 1] *= -1
+        return vec.dot(self.rotation)
+
+    def align_gradient(self, grad):
+        """suitable for vector system attached to atoms"""
+
+        # sensible? TODO
+        #algrad = np.copy(grad)
+        #if self.mirror:
+        #    algrad[:, 1] *= -1
+        algrad = grad.dot(self.rotation)
+        algrad = algrad[self.atommap]
+
+        return algrad
+
+    def align_hessian(self, hess):
+        # UNTESTED
+
+        blocked_hess = qcdb.util.blockwise_expand(hess, (3, 3), False)
+        alhess = np.zeros_like(blocked_hess)
+
+        nat = blocked_hess.shape[0]
+        for iat in range(nat):
+            for jat in range(nat):
+                alhess[iat, jat] = (self.rotation.T).dot(blocked_hess[iat, jat].dot(self.rotation))
+
+        alhess = alhess[:, self.atommap]
```

Comment:
Should be able to do `alhess[self.atommap[:, None], self.atommap]`.

Reply:
Why yes, not that I knew that when I wrote it. Even cleaner, I think I can do `alhess[np.ix_(self.atommap, self.atommap)]`

Reply:
Works as well :)